### ISSUE TITLE
Can you add the implementation of the Boundary-weighted regression loss function?

### ISSUE THREAD

**Issue by DatDoc** [wangyunpengbio/DeepSDM#1]

**Comment by wangyunpengbio** [comment]
It is a boundary-weighted L2 loss.
```
with torch.no_grad():
    batch_mask_numpy = batch_mask.cpu().numpy()
    gt_dis = self.sdf_fn_boundary(batch_mask_numpy) # function name: compute_sdf_normalize
    contour_map = self.loss_fn_boundary(batch_mask_numpy) # function name: compute_boundary
    contour_map = contour_map - gt_dis
    gt_dis = torch.from_numpy(gt_dis).float().to(self.device,non_blocking=True)
    contour_map = torch.from_numpy(contour_map).float().to(self.device,non_blocking=True)

distancemap_predict_tanh = torch.tanh(distancemap_predict)
# compute L2 loss between SDF Prediction and GT_SDF
loss_multihead_sdf_l2 = (distancemap_predict_tanh - gt_dis) ** 2 # mse loss
loss_multihead_sdf_l2 = contour_map * loss_multihead_sdf_l2 # contour_map work as the weight map
loss_multihead_sdf_l2 = loss_multihead_sdf_l2.mean()
loss_distance = loss_multihead_sdf_l2
loss = loss + loss_multihead_sdf_l2
```
```
from scipy.ndimage import distance_transform_edt as distance
def compute_sdf_normalize(img_gt):
    """
    compute the normalized signed distance map of binary mask
    input: segmentation, shape = (batch_size, x, y, z)
    output: the Signed Distance Map (SDM)
    sdf(x) = 0; x in segmentation boundary
             -inf|x-y|; x in segmentation
             +inf|x-y|; x out of segmentation
    normalize sdf to [-1, 1]
    """

    img_gt = img_gt.astype(np.uint8)

    normalized_sdf = np.zeros(img_gt.shape)

    for b in range(img_gt.shape[0]): # batch size
        for c in range(img_gt.shape[1]):
            posmask = img_gt[b].astype(np.bool)
            if posmask.any():
                negmask = ~posmask
                posdis = distance(posmask)
                negdis = distance(negmask)
                sdf = (negdis-np.min(negdis))/(np.max(negdis)-np.min(negdis)) - (posdis-np.min(posdis))/(np.max(posdis)-np.min(posdis))
                normalized_sdf[b][c] = sdf

    return normalized_sdf
```
```
def compute_boundary(img_gt):
    """
    compute the contour of binary mask
    input: segmentation, shape = (batch_size, x, y, z)
    output: the Contour Map
    Contour(x) = 2; x in segmentation boundary
             1; others
    """

    img_gt = img_gt.astype(np.uint8)

    contour_map = np.ones(img_gt.shape)

    for b in range(img_gt.shape[0]): # batch size
        for c in range(img_gt.shape[1]):
            posmask = img_gt[b].astype(np.bool)
            if posmask.any():
                boundary = skimage_seg.find_boundaries(posmask, mode='inner').astype(np.uint8) * 1
                contour_map[b][c] = contour_map[b][c] + boundary

    return contour_map
```
